File: week1_programming_challenges/2_maximum_pairwise_product/max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                (long long) numbers[first] * numbers[second]);
        }
    }

    return max_product;
}


long long MaxPairwiseProductFast(const std::vector<int>& numbers) {
    long long max_product = 0;
    int n = numbers.size();
    int max_index_1 = 0;
    int max_index_2 = 0;

    for (int first = 0; first < n; ++first) {
        if (numbers[first] > numbers[max_index_1]) {
            max_index_1 = first;
        }
    }

    for (int second = 0; second < n; ++second) {
        // std::cout << "second " << second << "\n";
        // std::cout << "max_index_2 " << max_index_2 << "\n";
        // std::cout << "numbers[second] " << numbers[second] << "\n";
        // std::cout << "numbers[max_index_2] " << numbers[max_index_2] << "\n";
        // std::cout << "max_index_1 " << max_index_1 << "\n";
        // std::cout << "max_index_1 " << max_index_1 << "\n";
        // std::cout << (max_index_2 == 0 || numbers[second] > numbers[max_index_2]) << "\n";
        // std::cout << (second != max_index_1) << "\n";
        
        if ((max_index_2 == 0 || numbers[second] > numbers[max_index_2]) && (second != max_index_1)) {
            // std::cout << "INNER" << "\n";
            max_index_2 = second;
        }
    }

    max_product = (long long) numbers[max_index_1] * numbers[max_index_2];

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProductFast(numbers) << "\n";
    return 0;
}

File: week1_programming_challenges/2_maximum_pairwise_product/MaxPairwiseProduct.java
import java.util.*;
import java.io.*;

public class MaxPairwiseProduct {
    static long getMaxPairwiseProduct(int[] numbers) {
        int n = numbers.length;
        int max_index1 = -1;
        int max_index2 = -1;

        for (int first = 0; first < n; first++) {
            if ((max_index1 == -1) || (numbers[first] > numbers[max_index1])) {
                max_index1 = first;
            }
        }

        for (int second = 0; second < n; ++second) {
            if (((max_index2 == -1) || (numbers[second] > numbers[max_index2])) && (second != max_index1)) {
                max_index2 = second;
            }
        }
        
        long product = (long) numbers[max_index1] * numbers[max_index2];
        return product;
    }

    public static void main(String[] args) {
        FastScanner scanner = new FastScanner(System.in);
        int n = scanner.nextInt();
        int[] numbers = new int[n];
        for (int i = 0; i < n; i++) {
            numbers[i] = scanner.nextInt();
        }
        System.out.println(getMaxPairwiseProduct(numbers));
    }

    static class FastScanner {
        BufferedReader br;
        StringTokenizer st;

        FastScanner(InputStream stream) {
            try {
                br = new BufferedReader(new
                    InputStreamReader(stream));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        String next() {
            while (st == null || !st.hasMoreTokens()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }
    }

}
